core/ClassInfo.php

```diff
@@ -52,17 +52,20 @@ public static function hasTable($class) {
 			return false;
 		}
 	}
-	
+
 	public static function reset_db_cache() {
 		self::$_cache_all_tables = null;
 		self::$_cache_ancestry = array();
 	}
-	
+
 	/**
 	 * Returns the manifest of all classes which are present in the database.
 	 * @param string $class Class name to check enum values for ClassName field
 	 */
 	public static function getValidSubClasses($class = 'SiteTree', $includeUnbacked = false) {
+		if (self::hasTable($class)) {
+			$class = self::$_cache_all_tables[strtolower($class)];
```

Comment:
Singletons should only be hard to do the first time they're called.

There's nothing wrong with strtolower... it's the hasTable that invokes the database and is my concern. On the other hand, we probably shouldn't be calling db queries on tables that may not exist, so maybe that should be there anyway.

Let's complain about another thing. :) I'm not sure that you should rely on the $_cache_all_tables private static outside of the method that generates the cache, since it makes it harder to read.

How about you use the suggested `normalise_class_name` method here instead?